R/connect_email.R

```diff
@@ -0,0 +1,173 @@
+#' While the output is invisible, any errors related to rendering will be
+#' visible to the author.
+#'
+#' @param email A rendered email message. Normally, we'd want to use an
+#'   associated .Rmd file with the `blastula::blastula_email` R Markdown output
+#'   format in [render_connect_email()] call (where its `input` is the email
+#'   .Rmd file).
+#' @param subject An option to specify the the email subject while attaching the
+#'   email object.
+#' @param attachments A vector of attachments for the Connect email. These files
+#'   can be any of those deployed when publishing to RStudio Connect, and, any
+#'   generated files (via R Markdown rendering).
+#' @param attach_output Should the rendered output of the main R Markdown
+#'   document be included as an email attachment? By default, this is `FALSE`.
+#'   If `TRUE` the main R Markdown document will be attached first (before any
+#'   files specified in `attachments`) and the filename will be preserved.
+#' @param text Instead of using a rendered email document through the `email`
+#'   option, we can use plain text here. If any text is provided here, input to
+#'   `email` is ignored.
+#' @param preview Should the email message display it's own preview window? If
+#'   `TRUE` (the default), the rendered email message will be shown in the
+#'   default browser.
+#'
+#' @export
+attach_connect_email <- function(email = NULL,
+                                 subject = NULL,
+                                 attachments = NULL,
+                                 attach_output = FALSE,
+                                 text = NULL,
+                                 preview = TRUE) {
+
+  if (!is.null(email) && !inherits(email, "email_message")) {
+    stop("A blastula email message object must be supplied.",
+         call. = FALSE)
+  }
+
+  if (is.null(email) && is.null(text)) {
+    stop("Either a blastula email message object or `text` must be provided",
+         call. = FALSE)
+  }
+
+  # If both an `email` object and `text` are provided, nullify
+  # the `text` and give preference to the HTML email
+  if (!is.null(email) && !is.null(text)) {
+    text <- NULL
+  }
+
+  if (!is.null(email)) {
+
+    if (is.na(Sys.getenv("RSC_REPORT_NAME", unset = NA))) {
+
+      # warning("connect_email() has no effect outside of RStudio Connect")
+
+      if (isTRUE(preview)) {
+
+        html_file <- tempfile(fileext = ".html")
+        html <- email$html_html
+
+        msg <- create_rmd_preview_message(subject = subject)
+
+        html <-
+          sub(
+            "(<body(?!\\w)[^>]*>)", paste0("\\1", msg),
+            html, perl = TRUE, ignore.case = TRUE
+          )
+
+        writeLines(html, html_file)
+        utils::browseURL(html_file)
+
+        # This sleep is necessary because knitting usually happens in a separate
+        # process, and when that process terminates the temp file will be deleted
+        Sys.sleep(5)
```

Comment:
This is a huge bummer, maybe there's another place we can put the HTML file?

Reply:
Couple of ideas:

* `{reportname}_files` or whatever that directory is that knitr creates
* If `self_contained` (or maybe always), we could call it `{reportname}_email.html` and put it in the same directory as the output

Reply:
Couldn't we use a character string in `utils::browseURL()`? I'll test that out.

Reply:
Scratch that. Needs to located somewhere.